pkpdapp/pkpdapp/models/inference.py

```diff
@@ -0,0 +1,356 @@
+        choices=SamplingAlgorithm.choices,
+        default=SamplingAlgorithm.HB,
+        help_text='sampling algorithm to use for inference')
+
+    class OptimisationAlgorithm(models.TextChoices):
+        CMAES = 'CMAES', 'CMAES'
+        XNES = 'XNES', 'XNES'
+        SNES = 'SNES', 'SNES'
+        PSO = 'PSO', 'PSO'
+        NM = 'NM', 'Nelder-Mead'
+
+    optimisation_algorithm = models.CharField(
+        choices=OptimisationAlgorithm.choices,
+        default=OptimisationAlgorithm.CMAES,
+        help_text='optimisation algorithm to use for inference')
+
+    number_of_iterations = models.IntegerField(
+        help_text='number of iterations'
+    )
+
+    time_elapsed = models.IntegerField(
+        default=0,
+        help_text='Elapsed run time for inference in seconds'
+    )
+
+    # potentially for optimisation too (as in number of starting points)
+    number_of_chains = models.IntegerField(
+        help_text='number of chains'
+    )
+
+    number_of_function_evals = models.IntegerField(
+        help_text='number of function evaluations'
+    )
+
+
+class StoredPharmacodynamicModel(PharmacodynamicModel):
+    """
+    Stored PD model.
+    """
+
+
+class StoredPharmacokineticModel(PharmacokineticModel):
+    """
+    Stored PK model.
+    """
+
+
+class DosedPharmacokineticModel(DosedPharmacokineticModel):
+    """
+    Stored dosed PK model.
+    """
+
+
+class ObjectiveFunction(Variable):
+    """
+    Abstract model class for objective functions.
+    """
+    biomarker_type = models.OnetoOne(
+        BiomarkerType,
+        on_delete=models.CASCADE,
+        blank=True, null=False)
+    pd_model = models.ForeignKey(
+        StoredPharmacodynamicModel,
+        blank=True, null=True,
+        on_delete=models.CASCADE,
+        help_text='pharmacodynamic model'
+    )
+    pk_model = models.ForeignKey(
+        StoredPharmacokineticModel,
+        blank=True, null=True,
+        on_delete=models.CASCADE,
+        help_text='pharmacokinetic model'
+    )
+    dosed_pk_model = models.ForeignKey(
+        StoredDosedPharmacokineticModel,
+        blank=True, null=True,
+        on_delete=models.CASCADE,
+        help_text='dosed pharmacokinetic model'
+    )
+    inference = models.ForeignKey(Inference, blank=True, null=False)
+
+
+class LogLikelihoodNormal(ObjectiveFunction):
+    """
+    Model for the normal log-likelihood.
+    """
+    sd = models.FloatField(
+        help_text='sd of normal prior distribution.'
+    )
+
+
+class LogLikelihoodLogNormal(ObjectiveFunction):
+    """
+    Model for the log-normal log-likelihood.
+    """
+    sigma = models.FloatField(
+        help_text='sigma of log-normal prior distribution.'
+    )
+
+
+class SumOfSquaredErrorsScoreFunction(ObjectiveFunction):
+    """
+    Model for sum of squared errors score function.
+    """
+
+
+class InferenceParameterProperties(Variable):
+    """
+    Abstract model that points to inference model.
+    """
+    pd_model = models.ForeignKey(
```

Comment:
same here. You might need an abstract base containing a lot of `Variable`. Then `Variable` inherits from that and puts in a bunch of fields pointing to pd_model, pk_model etc. Then have a `StoredVariable` that also inherits from the same thing and put a bunch of fields pointing to stored_pd_model, stored_pk_model etc, and an inference object. Then the priors and log likelihoods etc inherit from that

can define an abstract django model like so:
```python
class BaseMessage(models.Model):
    is_public = models.BooleanField(default=False)
    # some more fields...

    class Meta:
        abstract = True
```